waiter/src/waiter/instance_tracker.clj

```diff
@@ -0,0 +1,99 @@
+(ns waiter.instance-tracker
+  (:require [clojure.core.async :as async]
+            [clojure.set :as set]
+            [clojure.tools.logging :as log]
+            [metrics.timers :as timers]
+            [waiter.correlation-id :as cid]
+            [waiter.metrics :as metrics]
+            [waiter.util.utils :as utils]))
+
+; Events are being handled by all routers in a cluster for resiliency
+(defprotocol InstanceEventHandler
+
+  (handle-instances-event! [this instances]
+    "handles the instances event")
+
+  (state [this include-flags]
+    "returns the state of the handler:
+    {:last-error-time}"))
+
+(defrecord DefaultInstanceFailureHandler [clock
+                                          handler-state]
+
+  InstanceEventHandler
+
+  (handle-instances-event! [this instances]
```

Comment:
`instances` should instead be a map and we can currently populate `:new-failed-instances` and `:all-failed-instances`

Reply:
why do we need to provide all failed instances to the handler? The handler will only process the newly failed instances